### PR TITLE
fix: handle failed validations correctly [DET-3838]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Previously, a change was made that prevented completed validation workloads from completing in the experiment actor if they were missing validation metrics. This doesn't work because validation workloads that fail are intentionally missing metrics. Instead we check if the exit reason is nil before trying to compute the best validation.

## Test Plan
- [x] run a trial that fails in validation
- [x] add a unit test for a trial that fails in validation

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: e2e_tests/tests/fixtures/no_op/fail_on_first_validation.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+description: noop_single
```
Comment (rb-determined-ai) [inline e2e_tests/tests/fixtures/no_op/fail_on_first_validation.yaml:1]: Can we  reuse some other config file and use a temp config to set `fail_on_first_validation` in the code of the test?

File: e2e_tests/tests/test_system.py [hunk 2]
```diff
@@ -471,3 +471,11 @@ def test_pytorch_parallel() -> None:
...
         config, conf.official_examples_path("trial/mnist_pytorch"), 1
     )
+
+@pytest.mark.e2e_cpu  # type: ignore
+def test_fail_on_first_validation() -> None:
+    exp_id = exp.run_failure_test(
+        conf.fixtures_path("no_op/fail_on_first_validation.yaml"),
+        conf.fixtures_path("no_op"),
+        "failed on first validation",
+    )
```
Comment (rb-determined-ai) [inline e2e_tests/tests/test_system.py:481]: Should we also verify that the trial is in an errored state?  Or is that part of the run_failure_test?

Comment (stoksc) [reply e2e_tests/tests/test_system.py:482]: it checks this, just a little indirectly - there must be a trial in the 'ERRORED' state with this log line, so the check for the log line will fail if there is no errored trial.

File: e2e_tests/tests/fixtures/no_op/fail_on_first_validation.yaml (hunk 1 above)
Comment (stoksc) [reply e2e_tests/tests/fixtures/no_op/fail_on_first_validation.yaml]: yeah, added `run_failure_test_with_temp_config` to allow this.

File: e2e_tests/tests/fixtures/no_op/model_def.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.chaos_probability_train = self.env.hparams.get("chaos_probability_train")
         self.chaos_probability_validate = self.env.hparams.get("chaos_probability_validate")
         self.chaos_probability_checkpoint = self.env.hparams.get("chaos_probability_checkpoint")
+        self.fail_on_first_validation = self.env.hparams.get("fail_on_first_validation", "")
```
Comment (rb-determined-ai) [inline e2e_tests/tests/fixtures/no_op/model_def.py:38]: default to `False`?

File: e2e_tests/tests/test_system.py [hunk 4]
```diff
@@ -471,3 +471,13 @@ def test_pytorch_parallel() -> None:
     exp.run_basic_test_with_temp_config(
         config, conf.official_examples_path("trial/mnist_pytorch"), 1
     )
+
+
+@pytest.mark.e2e_cpu  # type: ignore
+def test_fail_on_first_validation() -> None:
+    error_log = "failed on first validation"
```
Comment (rb-determined-ai) [inline e2e_tests/tests/test_system.py:478]: not sure what this variable is actually for?

Comment (stoksc) [reply e2e_tests/tests/test_system.py:478]: i forgot to change it, it's what is raised in the model def.

**Pull request merged**